### DISCUSSION TITLE
CLI v3 Breaking Changes

### DISCUSSION THREAD

**Discussion by Spacellary** [nikhilbadyal/docker-py-revanced#323] [General]
### Context:
> (...) I just noticed when using the new pre-release versions of the ReVanced/revanced-cli locally, `-a` is no longer being used to specify the target APK for patching, instead the `patch` command is used, it used to be:
> 
> `java -jar cli.jar -a duolingo.apk -b patches.jar...
`> 
> ...but now it is:
> 
> `java -jar cli.jar patch duolingo.apk -b patches.jar...`
> 
> This is gonna cause issues when patching with the official ReVanced CLI in their next stable ReVanced/revanced-cli release if it reflects their pre-release changes so far. (...)

Making a separate discussion for this specific matter as it's not necessaraly related to `app support`.

As discussed I can only wish they'd keep support for both instead of flat out replacing it, and it seems like a mistake as `-a` isn't being used by anything else since it's disappearance, it was just removed, so it could just have been an oversight.
___

### @IMXEren, @nikhilbadyal:

I am opening this discussion in hopes that we can consider the possibility to open an issue -- initially in **ReVanced/recanced-cli** itself -- and the best approach to do so with sufficient motive to provide convincing reasoning to keep `-a` as a short alias for `patch` before the next stable version of ReVanced/recanced-cli is released, avoiding a major headache for a minor change.

**Comment by IMXEren** [comment]
Great workaround you discovered.

Now, the ball is in the dev's court.

**Reply by Spacellary** [reply]
I'd like to avoid giving Nik more work over this than necessary, haha.

So actually I'd like to propose opening an issue initially in **ReVanced/recanced-cli** itself, reasoning to the team why they should keep the `-a` shorhand alongside `patch`. It's even unused. `-a` for source APK made perfect sense.

This dicussion will serve for that too among other things.

**Reply by Spacellary** [reply]
But yeah, one possible solution would be to run a *check*/test run command with `-h` with the supplied CLI, and check for the returned output pattern, if it matches the old structure, then use `-a` in the command, if not then use `patch` instead.

Or a manual toggle, as another option to hold things up. "PATCH_ARGUMENT_TYPE=`-a`/`patch`" until the method above can be completed.

**Reply by nikhilbadyal** [reply]
I looked into the new implementation. They basically reimplemented cli. Earlier `a`, `b` used to be individual commands. Now they categorized the command. `patch` is now a main command with `b` and all others as sub commands.
Not sure if they'll agree to keep `a` still there.

We can try though.

**Comment by Spacellary** [comment]
### Quick comparison.

"Old" `-h` output:

![WindowsTerminal_JLTCo1hbmW](https://github.com/nikhilbadyal/docker-py-revanced/assets/39604793/0f12dfdc-79f1-4849-82d4-2425cfedc39e)

<br>

"New" `-h` output:

![WindowsTerminal_WhehsrdFgG](https://github.com/nikhilbadyal/docker-py-revanced/assets/39604793/24192914-d20a-48d3-9487-e3e768fe69ff)

___

### Running `java -jar .\c.jar patch` on...

#### "Old" (first line):
```
Unmatched argument at index 0: 'patch'
```

#### "New" (first line):
```
Missing required options and parameters: '--patch-bundle=<patchBundles>', '--out=<outputFilePath>', '<apk>'
```

**Reply by nikhilbadyal** [reply]
It's going to be ugly for maintenance. Running `help` then somehow extracting the correct option.

**Reply by nikhilbadyal** [reply]
If we end up doing this. We can just hard code the logic. 
- If the `-v` (Version) > 3 use `patch` else use `a`. This look easier to maintain.

**Reply by Spacellary** [reply]
What about this?

> ### Running `java -jar .\c.jar patch` on...
> 
> #### "Old" (first line):
> ```
> Unmatched argument at index 0: 'patch'
> ```
> 
> #### "New" (first line):
> ```
> Missing required options and parameters: '--patch-bundle=<patchBundles>', '--out=<outputFilePath>', '<apk>'
> ```

**Reply by nikhilbadyal** [reply]
Using Version looks easy to me.

**Reply by Spacellary** [reply]
@nikhilbadyal You're right, I was checking that and it's the way to go then. Way cleaner, haha. Well, ideally they'd keep `-a` but since you said it's a complete rewrite it's important to be expecting them just *not* to.

**Reply by nikhilbadyal** [reply]
<img width="431" alt="image" src="https://github.com/nikhilbadyal/docker-py-revanced/assets/59223300/46a5ba4b-76aa-4ad5-96ad-af2365459d55">

<br>Can you check if current cli shows version ? Mine just says `unknown`.
@Spacellary @IMXEren

**Reply by Spacellary** [reply]
@nikhilbadyal Same here, using the current CLI...

![WindowsTerminal_0JgtVt9ysA](https://github.com/nikhilbadyal/docker-py-revanced/assets/39604793/3ab3083d-5033-4974-b4ea-4bfbebd437e7)

**Reply by nikhilbadyal** [reply]
Will do a blind check then. if `-v` contains 3 use `patch` else use `a`

**Reply by Spacellary** [reply]
Sounds good.
If --version ≤ 3 _then_ `patch`, _else_ `-a`.

Thanks for being proactive on this one by the way, fixing the issue before it even comes up is a good call.

It can be tested by linking to the pre-release tag. Another great functionality.

**Reply by nikhilbadyal** [reply]
Are you able to patch any app with the new v3 cli ? @Spacellary 

<img width="988" alt="image" src="https://github.com/nikhilbadyal/docker-py-revanced/assets/59223300/34770f59-0a8b-4864-99cc-9159cf88869d">

**Reply by Spacellary** [reply]
@nikhilbadyal Java doing its best~

I was able to patch Duolingo using the CLI v3. I'll try YT and YTM.

**Reply by nikhilbadyal** [reply]
Which CLI version and Patch version ?

**Reply by Spacellary** [reply]
@nikhilbadyal Wait, actually I was testing a pre-release version, so the Stable release (ReVanced CLI v3.0.0) has the same issue as you're facing.

```
INFORMA├ç├òES: Setting patch options
Exception in thread "main" java.lang.NoClassDefFoundError: app/revanced/patcher/patch/PatchResultError

        at java.base/java.lang.Class.getDeclaredFields0(Native Method)
        at java.base/java.lang.Class.privateGetDeclaredFields(Class.java:3297)
        at java.base/java.lang.Class.getDeclaredField(Class.java:2608)
        at kotlin.reflect.jvm.internal.KClassImpl$Data$objectInstance$2.invoke(KClassImpl.kt:113)
        at kotlin.reflect.jvm.internal.ReflectProperties$LazyVal.invoke(ReflectProperties.java:63)
        at kotlin.reflect.jvm.internal.ReflectProperties$Val.getValue(ReflectProperties.java:32)
        at kotlin.reflect.jvm.internal.KClassImpl$Data.getObjectInstance(KClassImpl.kt:108)
        at kotlin.reflect.jvm.internal.KClassImpl.getObjectInstance(KClassImpl.kt:242)
        at kotlin.reflect.full.KClasses.getCompanionObjectInstance(KClasses.kt:57)
        at app.revanced.patcher.extensions.PatchExtensions.getOptions(PatchExtensions.kt:60)
        at app.revanced.utils.Options.setOptions(Options.kt:60)
        at app.revanced.utils.Options.setOptions(Options.kt:89)
        at app.revanced.cli.command.PatchCommand.run(PatchCommand.kt:163)
        at picocli.CommandLine.executeUserObject(CommandLine.java:2026)
        at picocli.CommandLine.access$1500(CommandLine.java:148)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
        at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
        at picocli.CommandLine.execute(CommandLine.java:2170)
        at app.revanced.cli.command.MainCommandKt.main(MainCommand.kt:40)
Caused by: java.lang.ClassNotFoundException: app.revanced.patcher.patch.PatchResultError
        at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:445)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:592)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
        ... 22 more
```

**Reply by Spacellary** [reply]
@nikhilbadyal Seems like an issue with patches?

Latest CLI v3 + Integrations + Patches [v2.188.0-dev.25](https://github.com/ReVanced/revanced-patches/releases/tag/v2.188.0-dev.25) worked. And that's what I had before, for patching Duolingo, tested again now and it works.


revanced-cli-3.0.0
revanced-integrations-0.116.2
revanced-patches-[v2.188.0](https://github.com/ReVanced/revanced-patches/releases/tag/v2.188.0)

So using 2.188.0 as they said in your Issue works. Huh, yea that's weird, up until now the latest stable resources were all compatible, very confusing. This should sort itself out as they release a new stable CLI version.

For now it's just necessary to specify `GLOBAL_PATCHES_DL=https://github.com/ReVanced/revanced-patches/releases/tag/v2.188.0` for those that can't wait.

**Comment by nikhilbadyal** [comment]
Handled [now](https://github.com/nikhilbadyal/docker-py-revanced/commit/10d91cc0f26f62db4a80402acdbb853fab68b014) and [works](https://github.com/nikhilbadyal/docker-py-revanced/actions/runs/5990337055/job/16259300205#step:6:355) fine. You can test and close the discussion.

**Reply by nikhilbadyal** [reply]
@Spacellary

**Reply by Spacellary** [reply]
I'll give it a go once I am home in the evening. 👍

**Reply by Spacellary** [reply]
@nikhilbadyal Sorry for the delay.

Patching with v3 is all good, [working perfectly here](https://github.com/Spacellary/ReVanced-Extended-Automated-Builder/actions/runs/6007640340/job/16294062394#step:6:77), patching Duolingo with APKMirror link.

CLi v2 patching [also working as expected](https://github.com/Spacellary/ReVanced-Extended-Automated-Builder/actions/runs/6007640340/job/16294062394#step:6:77). Thanks for quick work on this one.

Closing discussion.

@IMXEren No more breaking anytime soon we hope xD

**Reply by IMXEren** [reply]
Hope so...still the dev would make me sync the newer changes 😅.

@nikhilbadyal thank you for your quick fixes.

**Comment by Spacellary** [comment]
*Patching with both v2 and v3 CLI working as expected now.*
